### ISSUE TITLE
[Question] loading pretrained model weights into a new model

### ISSUE THREAD

**Issue by WreckItTim** [DLR-RM/stable-baselines3#1414]
### ❓ Question

Hello,

I saw the previous post here https://github.com/DLR-RM/stable-baselines3/issues/543 with the corresponding paper and google collab notebooks. These helped for sure, thank you!

I am doing something slightly different, and rather than replacing the entire policy with a custom policy object, I want to just replace the nn model in a new SB3 model object - after doing some supervised learning with a surrogate model. To be specific... 

1. I have a custom torch.nn model with pretrained weights for both the actor and critic, and I want to create a new TD3 policy to use these models.
2. As the previously mentioned paper suggests, I want to collect some data into the replay buffer with the pretrained actor/critic before any online training occurs when fine-tuning.

Solution (psudeo)-code:

```python
n_episodes_b4 = 100 # first collect 100 episodes using pretrained networks
n_episodes_after = 1 # train every n_episodes_after episodes after collection
sb3model = stable_baselines3.TD3(..., learning_starts=0, train_freq=(n_episodes_b4, 'episode'))

# fetch pretrained actor and critic torch.nn models
pretrained_actor, pretrained_critics = load -> torch.nn.Sequential(...), [torch.nn.Sequential(...), ...]

# replace actor network with pretrained actor
sb3model.actor.mu = copy.deepcopy(pretrained_actor)
sb3model.actor_target.mu = copy.deepcopy(pretrained_actor)

# replace all critic networks with pretrained critics
for module in sb3model.critic.modules():
    del module
for module in sb3model.critic_target.modules():
    del module
for idx in range(len(sb3model.critic.q_networks)):
    q_net = copy.deepcopy(pretrained_critics[idx])
    sb3model.critic.q_networks[idx] = q_net
    sb3model.critic.add_module(f"qf{idx}", q_net)
    q_net_target = copy.deepcopy(pretrained_critics[idx])
    sb3model.critic_target.q_networks[idx] = q_net_target
    sb3model.critic_target.add_module(f"qf{idx}", q_net_target)
    
# use pretrained model to collect n_collection steps before training, not rando
sb3model.learn(env, ...)
#  then start training on normal schedule
env -> after 100 reset() calls then set:
    sb3model.train_freq = (n_episodes_after, 'episode')
    sb3model._convert_train_freq()
```

Note that I set the actual network here rather than using load_state_dict() because I will be using custom network layers as well. So I do not want to just set weights. I want to replace the modules with my own sequential model, which also has pretrained weights.

I am a little nervous when playing with such a big repo as SB3, and hoping I did not miss anything that may break training. Everything compiles and looks right, but are there any logic errors or other issues I am missing? Thank you so much!

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] If code there is, it is minimal and working
- [X] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by BeFranke** [comment]
I had the same problem, and wrote a wrapper subclass to stable_baselines3.common.policies.ActorCriticPolicy to achieve this:

`from stable_baselines3.common.policies import ActorCriticPolicy
from stable_baselines3.common.type_aliases import Schedule
from stable_baselines3.common.torch_layers import FlattenExtractor
from stable_baselines3.common.distributions import (
    BernoulliDistribution,
    CategoricalDistribution,
    DiagGaussianDistribution,
    Distribution,
    MultiCategoricalDistribution,
    StateDependentNoiseDistribution,
    make_proba_distribution,
)

from gymnasium import spaces
from torch import nn
import torch as th

from typing import Type, Dict, Optional, Union, Any, List, Tuple

class ModelRole(Enum):
    POLICY = auto()
    VALUE = auto()

class TorchWrapperActorCriticPolicy(ActorCriticPolicy):
    def __init__(self, 
                 observation_space: spaces.Space, 
                 action_space: spaces.Space, 
                 lr_schedule: Schedule, 
                 activation_fn: Type[nn.Module] = nn.Tanh, 
                 ortho_init: bool = True, 
                 use_sde: bool = False, 
                 log_std_init: float = 0, 
                 full_std: bool = True, 
                 use_expln: bool = False, 
                 squash_output: bool = False, 
                 optimizer_class: Type[th.optim.Optimizer] = th.optim.Adam, 
                 optimizer_kwargs: Optional[Dict[str, Any]] = None,
                 torch_policy_network: Optional[nn.Module] = None,
                 torch_value_network: Optional[nn.Module] = None):
        
        super().__init__(observation_space, action_space, lr_schedule, None, activation_fn, ortho_init, 
                         use_sde, log_std_init, full_std, use_expln, squash_output, FlattenExtractor, 
                         None, True, True, optimizer_class, optimizer_kwargs)
        self._policy_network = torch_policy_network
        self._value_network = torch_value_network
        
        self._patch(self.lr_schedule)
        
    def _build(self, lr_schedule: Schedule) -> None:
        """
        We do not want to use this, since most initialization depends on the presence of the policy and value networks.
        However, these can only be set after super()__init__(), as this call also initializes the torch.nn.Module.
        Therefore, the fuctionality is moved to self._patch.
        """
        self.lr_schedule = lr_schedule
        
        
    def forward(self, obs: th.Tensor, deterministic: bool = False) -> Tuple[th.Tensor, th.Tensor, th.Tensor]:
        """
        Forward pass in all the networks (actor and critic)

        :param obs: Observation
        :param deterministic: Whether to sample or use deterministic actions
        :return: action, value and log probability of the action
        """
        features = self.extract_features(obs)
        actions = self._policy_network(features)
        values = self._value_network(features)
        distribution = self._get_action_dist_from_latent(actions)
        actions = distribution.get_actions(deterministic=deterministic)
        log_prob = distribution.log_prob(actions)
        actions = actions.reshape((-1, *self.action_space.shape))
        return actions, values, log_prob
    
    def _patch(self, lr_schedule: Schedule) -> None:
        # ensure that the functionality we want to replace is actually removed
        del self.pi_features_extractor
        del self.vf_features_extractor
        
        self.log_std = nn.Parameter(th.ones(self.action_dist.action_dim) * self.log_std_init, requires_grad=True)
        # Setup optimizer with initial learning rate
        self.optimizer = self.optimizer_class(self.parameters(), lr=lr_schedule(1), **self.optimizer_kwargs)
        
        if self.ortho_init:
            for s, network, role in zip(["policy", "value"], [self._policy_network, self._value_network], [ModelRole.POLICY, ModelRole.VALUE]):
                try:
                    network.ortho_init(role)
                except AttributeError:
                    print(f"{s}-network does not have a method ortho_init(). Skipping weight initialization!")
        
    def _get_action_dist_from_latent(self, mean_actions: th.Tensor) -> Distribution:
        """
        Retrieve action distribution given the latent codes.

        :param mean_actions: output of self._policy_network
        :return: Action distribution
        """
        if isinstance(self.action_dist, DiagGaussianDistribution):
            return self.action_dist.proba_distribution(mean_actions, self.log_std)
        elif isinstance(self.action_dist, CategoricalDistribution):
            # Here mean_actions are the logits before the softmax
            return self.action_dist.proba_distribution(action_logits=mean_actions)
        elif isinstance(self.action_dist, MultiCategoricalDistribution):
            # Here mean_actions are the flattened logits
            return self.action_dist.proba_distribution(action_logits=mean_actions)
        elif isinstance(self.action_dist, BernoulliDistribution):
            # Here mean_actions are the logits (before rounding to get the binary actions)
            return self.action_dist.proba_distribution(action_logits=mean_actions)
        # elif isinstance(self.action_dist, StateDependentNoiseDistribution):
        #    return self.action_dist.proba_distribution(mean_actions, self.log_std, latent_pi)
        else:
            raise ValueError("Invalid action distribution")
        
    def get_distribution(self, obs: th.Tensor) -> Distribution:
        """
        Get the current policy distribution given the observations.

        :param obs:
        :return: the action distribution.
        """
        features = super().extract_features(obs)
        mean_actions = self._policy_network(features)
        return self._get_action_dist_from_latent(mean_actions)
    
    def predict_values(self, obs: th.Tensor) -> th.Tensor:
        """
        Get the estimated values according to the current policy given the observations.

        :param obs: Observation
        :return: the estimated values.
        """
        features = super().extract_features(obs)
        return self._value_network(features)
    
    def evaluate_actions(self, obs: th.Tensor, actions: th.Tensor) -> Tuple[th.Tensor, th.Tensor, Optional[th.Tensor]]:
        """
        Evaluate actions according to the current policy,
        given the observations.

        :param obs: Observation
        :param actions: Actions
        :return: estimated value, log likelihood of taking those actions
            and entropy of the action distribution.
        """
        # Preprocess the observation if needed
        features = self.extract_features(obs)
        mean_actions = self._policy_network(features)
        
        distribution = self._get_action_dist_from_latent(mean_actions)
        log_prob = distribution.log_prob(actions)
        values = self._value_network(features)
        entropy = distribution.entropy()
        return values, log_prob, entropy
    
`

Now, I am pretty new to stable baselines 3 and do not claim that this is a good way to do it, but from some limited experiments I did this at least achieves a comparable performance to the sb3-native solution when given the same networks and hyperparameters.

With this code, you can setup your policy and value networks as torch MLPs and then pass them like this (example for PPO):

`model = PPO(env=train_env, policy=TorchWrapperActorCriticPolicy, policy_kwargs={
            "torch_policy_network": <put your custom policy network here>,
            "torch_value_network": <put your custom policy network here>
        })`


You should then be able to freely re-assign model.policy._policy_network and model.policy._value_network. 
Feedback to this solution is welcome!

**Comment by araffin** [comment]
> I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)

https://stable-baselines3.readthedocs.io/en/master/guide/examples.html#accessing-and-modifying-model-parameters

>I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo

Probably a duplicate of https://github.com/DLR-RM/stable-baselines3/issues/1411

**Comment by WreckItTim** [comment]
> > I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
> 
> https://stable-baselines3.readthedocs.io/en/master/guide/examples.html#accessing-and-modifying-model-parameters
> 
> TL;DR: you can use `set_parameters` or `load_state_dict()` (you will probably need to rename some keys)
> 
> > I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
> 
> Probably a duplicate of #1411

Slightly a duplicate, the difference being I am also trying to collect a replay buffer using the already trained model before online training starts. As opposed to randomly sampling the action space before online training starts.

**Comment by araffin** [comment]
The code that update the replay buffer is there: https://github.com/DLR-RM/stable-baselines3/blob/5a70af8abddfc96eac3911e69b20f19f5220947c/stable_baselines3/common/off_policy_algorithm.py#L416

In the examples, we have one about saving/loading a replay buffer.

**Comment by WreckItTim** [comment]
@araffin Perfect! This might be better than the hacky solution I was using in original post. Thank you for the responses =)

@araffin Yes this is similar to what I want to do! Thank you. I was trying to avoid making a custom class, because you never know for sure if there are any potential logic errors, or other steps, you are skipping/overwriting when working with a large, external repo. I think it's a better solution than overriding/hacking the built-in classes though.
